tuning/utils/preprocessing_utils.py

```diff
@@ -0,0 +1,329 @@
+# Copyright The FMS HF Tuning Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# Standard
+from typing import Any, Callable, Dict, Optional
+import json
+
+# Third Party
+from datasets import Dataset
+from transformers import AutoTokenizer, DataCollatorForSeq2Seq
+from trl import DataCollatorForCompletionOnlyLM
+import datasets
+
+# Local
+from tuning.config import configs
+
+
+def get_data_trainer_kwargs(
+    training_data_path: str,
+    validation_data_path: str,
+    packing: bool,
+    response_template: Optional[str],
+    max_sequence_length: int,
+    tokenizer: AutoTokenizer,
+    dataset_text_field: Optional[str],
+) -> Dict[str, Any]:
+    """Get trainer args related to data / processing. At the moment, this consists of:
+        - the training dataset
+        - the evaluation dataset
+        - the data collator
+        - Maybe a formatting a function [only for a special case for validation]
+    The result can be kwarg expanded into the trainer initialization.
+
+    Args:
+        training_data_path: str
+            Path to the training data.
+        validation_data_path: str
+            Path to the validation data.
+        packing: bool
+            Whether or not we should apply packing or not.
+        response_template: Optional[str]
+            Response template to be used for formatting by TRL.
+        max_sequence_length: int
+            Max sequence length to be used for sequence tokenization.
+        tokenizer: AutoTokenizer
+            Loaded tokenizer object to be used by the collator.
+        dataset_text_field: Optional[str]
+            Dataset text field fto be used for formatting by TRL.
+
+    Returns:
+        Dict[str, Any]
+            Data related kwargs to be used by the SFT Trainer.
+    """
+    data_collator = get_data_collator(
+        packing, dataset_text_field, response_template, max_sequence_length, tokenizer
+    )
+    eval_dataset = None
+    data_kwargs = {}
+    if isinstance(data_collator, DataCollatorForSeq2Seq):
+        # HACK: This function is never called, but is needed to sidestep TRL's internal validation.
+        data_kwargs["formatting_func"] = lambda x: x
```

Comment:
I was implementing a kind of goofy workaround for TRL's validation logic with wrapping the tokenizer to make `tokenize()` a loop when I realized that there is actually already logic for handling pretokenized data based on the columns of the passed dataset: https://github.com/huggingface/trl/blob/main/trl/trainer/sft_trainer.py#L477

So all we have to do to handle the tokenized data is pass a dummy formatting function, because `_prepare_dataset` will inspect the dataset and return immediately without ever calling the value passed here.

There is additionally an extra argument, `dataset_kwargs`, which is a dict that can pass `skip_prepare_dataset`, which is bool-valued. I think the correct behavior is to just check that value in TRL and skip the validation for the `packing=True` case, which is a small change that does not affect its API or change the supported data formats.

I have opened an issue / pull request in TRL to this effect: https://github.com/huggingface/trl/pull/1673